x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RemoteAccessAuthentication.java

```diff
@@ -0,0 +1,160 @@
+        this.authentication = authentication;
+        this.roleDescriptorsBytesList = roleDescriptorsBytesList;
+    }
+
+    public void writeToContext(final ThreadContext ctx) throws IOException {
+        ctx.putHeader(REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY, encode());
+    }
+
+    public static RemoteAccessAuthentication readFromContext(final ThreadContext ctx) throws IOException {
+        return decode(ctx.getHeader(REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY));
+    }
+
+    public Authentication getAuthentication() {
+        return authentication;
+    }
+
+    public List<RoleDescriptorsBytes> getRoleDescriptorsBytesList() {
+        return roleDescriptorsBytesList;
+    }
+
+    private static List<RoleDescriptorsBytes> toRoleDescriptorsBytesList(final RoleDescriptorsIntersection roleDescriptorsIntersection)
+        throws IOException {
+        // If we ever lift this restriction, we need to ensure that the serialization of each set of role descriptors to raw bytes is
+        // deterministic. We can do so by sorting the role descriptors before serializing.
+        assert roleDescriptorsIntersection.roleDescriptorsList().stream().noneMatch(rds -> rds.size() > 1)
+            : "sets with more than one role descriptor are not supported for remote access authentication";
+        final List<RoleDescriptorsBytes> roleDescriptorsBytesList = new ArrayList<>();
+        for (Set<RoleDescriptor> roleDescriptors : roleDescriptorsIntersection.roleDescriptorsList()) {
+            roleDescriptorsBytesList.add(RoleDescriptorsBytes.fromRoleDescriptors(roleDescriptors));
+        }
+        return roleDescriptorsBytesList;
+    }
+
+    private String encode() throws IOException {
+        final BytesStreamOutput out = new BytesStreamOutput();
+        out.setVersion(authentication.getEffectiveSubject().getVersion());
+        Version.writeVersion(authentication.getEffectiveSubject().getVersion(), out);
+        authentication.writeTo(out);
+        out.writeCollection(roleDescriptorsBytesList, StreamOutput::writeBytesReference);
+        return Base64.getEncoder().encodeToString(BytesReference.toBytes(out.bytes()));
+    }
+
+    private static RemoteAccessAuthentication decode(final String header) throws IOException {
+        Objects.requireNonNull(header);
+        final byte[] bytes = Base64.getDecoder().decode(header);
+        final StreamInput in = StreamInput.wrap(bytes);
+        final Version version = Version.readVersion(in);
+        in.setVersion(version);
+        final Authentication authentication = new Authentication(in);
+        final List<RoleDescriptorsBytes> roleDescriptorsBytesList = in.readImmutableList(RoleDescriptorsBytes::new);
+        return new RemoteAccessAuthentication(authentication, roleDescriptorsBytesList);
+    }
+
+    public static final class RoleDescriptorsBytes extends AbstractBytesReference {
```

Comment:
Extending `AbstractBytesReference` allows us to write these to [authentication metadata](https://github.com/elastic/elasticsearch/pull/92089/files#diff-fda6e1b9b70b8d40bd3770a1c2c07b22c2c4b6bd7b7b4ed076a38389b8b15b50R75-R77)

Reply:
I like this abstraction!

Do we not want to put the entire RemoteAccessAuthentication into the authentication metadata? This really is for the other PR. So we can discuss it in future over there.

Reply:
👍 the class is much cleaner now thanks to your suggestions

I'm thinking of putting only selected fields from `RemoteAccessAuthentication` into auth metadata (similar to how we handle API keys), but perhaps the whole instance also makes sense. Serialization might get tricky in that case, but either way let's revisit during the fulfilling cluster implementation.